### ISSUE TITLE
clang introduces undefined behavior into well-formed C++ program

### ISSUE THREAD

**Issue by gonzalobg** [llvm/llvm-project#62057]
This program is well-formed and the C++ standard guarantees that a thread that reaches `block` will not execute any subsequent instructions, since the thread performs an [execution step] on every loop iteration (the atomic operation)](http://eel.is/c++draft/intro.progress#3.3).

Therefore, when `example(nullptr)` is called, the C++ standard guarantees that a null pointer will never be dereferenced.

However, clang ([godbolt](https://clang.godbolt.org/z/Tn46qv9ce)) removes the loop, causing the program to dereference a `nullptr`, therefore introducing _undefined behavior_ into a well-formed program. My assessment is that clang is incorrectly not inserting an `llvm.sideeffect` intrinsic inside the loop.

```cpp
#include <atomic>

void block() {
    std::atomic<int> x = 0;
    while (true) {
        if (x.load(std::memory_order_relaxed) == 1)
            break;
    }
}

void example(int* data) {
    block();
    *data = 42;
}

int main() {
   example(nullptr);
   return 0;
}
```

**Comment by Endilll** [comment]
There's a related paper that is currently going through WG21: https://github.com/cplusplus/papers/issues/1496

**Comment by RalfJung** [comment]
@Endilll that paper is not really related. This issue is about loops which C++ already explicitly says are Well-Defined (because they contain an atomic operation). That paper would only make even more loops defined.

**Comment by shafik** [comment]
This does look like a bug, CC @AaronBallman @dwblaikie since I am not sure who the right person to pull in for this.

**Comment by AaronBallman** [comment]
CC @rjmccall @efriedma-quic as codegen code owners

I don't think LLVM or Clang documents what forward progress guarantees the main thread has (http://eel.is/c++draft/intro.progress#8), but I would imagine we want to provide concurrent forward progress guarantees (http://eel.is/c++draft/intro.progress#7) which would make this a bug.

**Comment by efriedma-quic** [comment]
The issue is that LLVM's notion of forward progress doesn't quite agree with the standard.  Specifically, it doesn't treat a constant-foldable atomic load as "forward progress": it eliminates the load without leaving any indication in the IR that an atomic operation was eliminated.

> These idioms are the only reliable way to block C++ threads forever

This is obviously not true.  You can, for example, wait on an std::condition_variable (which is also a lot more efficient because it also indicates to the OS that the thread is idle).

**Comment by gonzalobg** [comment]
> This is obviously not true. You can, for example, wait on an std::condition_variable (which is also a lot more efficient because it indicates to the OS that the thread is idle).

Should have clarified with "...on freestanding" (since that is what I'm targetting). Is there a way to block a thread forever on freestanding targets?

**Comment by rjmccall** [comment]
> The issue is that LLVM's notion of forward progress doesn't quite agree with the standard. Specifically, it doesn't treat a constant-foldable atomic load as "forward progress": it eliminates the load without leaving any indication in the IR that an atomic operation was eliminated.

Yeah, having this optimization leave an `llvm.sideeffect` behind seems like the most straightforward way to fix the issue, although it unfortunately ties IR even more to C/C++ semantics.

**Comment by rjmccall** [comment]
> Is there a way to block a thread forever on freestanding targets without resorting to inline asm?

The simplest way would be to loop forever loading or storing to a volatile global.  I believe LLVM understands that it isn't allowed to fold those even if it can prove that they're the only accesses.

**Comment by efriedma-quic** [comment]
For reference:

- The relevant standard paragraph here is actually [intro.progess]p1: "The implementation may assume that any thread will eventually do one of the following [...]".
- The relevant bit of LangRef: https://llvm.org/docs/LangRef.html#langref-mustprogress .

> it unfortunately ties IR even more to C/C++ semantics

The definition of "mustprogress" is already tied to C++ semantics, and I don't think there's any way around that.  Maybe we can figure out some way to avoid complicating programs that aren't using "mustprogress" markings.

> > This is obviously not true. You can, for example, wait on an std::condition_variable (which is also a lot more efficient because it indicates to the OS that the thread is idle).
> 
> Should have clarified with "...on freestanding" (since that is what I'm targetting).

If your environment has threads, it should provide some way for a thread to block without spinning the CPU, and you should probably use it.

Otherwise, a loop involving a volatile operation or `asm("");` should be reliable.    (You mention "very sub-optimal code", but if the CPU isn't doing any useful work anyway, I can't see how that matters.)  Or you can use -fno-finite-loops to turn off the relevant optimizations.

**Comment by gonzalobg** [comment]
> If your environment has threads, it should provide some way for a thread to block without spinning the CPU, and you should probably use it.

Both C++ and clang support environments with threads that don't provide this. 

> You mention "very sub-optimal code", but if the CPU isn't doing any useful work anyway, I can't see how that matters.

Both C++ and clang support environments with more than one hardware thread, on which a hardware thread spinning doing nothing is significantly better than spinning by bombarding the memory subsystem with useless requests that may slow down the progress of other threads.

> Or you can use -fno-finite-loops to turn off the relevant optimizations. Otherwise, a loop involving a volatile operation or asm(""); 

The `-fno-finite-loops` seems like a pretty big hammer with wide performance impact. Spinning on a volatile `asm("")` sounds reasonable.

**Comment by rjmccall** [comment]
> Both C++ and clang support environments with more than one hardware thread, on which a hardware thread spinning doing nothing is significantly better than spinning by bombarding the memory subsystem with useless requests that may slow down the progress of other threads.

Your idiom has the same impact on the memory subsystem as loading from a volatile global would.

**Comment by rjmccall** [comment]
I agree that it's easier to optimize your idiom.  I'm just saying that the compilers which aren't miscompiling it do not in fact optimize away the load, so the idiom's performance benefit over a `volatile` load is currently just theoretical.  Also, in practice both patterns will just hit L1 forever and never bother the rest of memory.

**Comment by gonzalobg** [comment]
@rjmccall 

> Also, in practice both patterns will just hit L1 forever and never bother the rest of memory.

This does not hold for many environments clang supports. For example, the NVPTX backend will spill the automatic variable to `local` memory and perform local memory accesses while spinning.
